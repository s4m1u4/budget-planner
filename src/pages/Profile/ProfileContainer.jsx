import Profile from "./Profile";
import { inject, observer } from "mobx-react";

export const ProfileContainer = inject(
  ({
    rootStore: {
      userStore: { getUserData, setNewUserData, isLoading },
    },
  }) => ({ getUserData, setNewUserData, isLoading })
)(observer(Profile));
