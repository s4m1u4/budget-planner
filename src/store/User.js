import { makeAutoObservable } from "mobx";

class User {
  login = false;

  constructor() {
    makeAutoObservable(this);
  }

  setUser() {
    this.login = true;
  }
}

export default new User();
