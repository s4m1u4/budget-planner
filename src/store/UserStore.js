import { makeAutoObservable } from "mobx";

class UserStore {
  isAuth = false;
  userData = {};

  constructor() {
    makeAutoObservable(this);
  }

  setIsAuth() {
    this.isAuth = true;
  }

  setSignupData(data) {
    this.userData = data;
  }
}

const userStore = new UserStore();

export default userStore;
