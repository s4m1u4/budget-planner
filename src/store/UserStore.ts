import { makeAutoObservable } from "mobx";
import { getToken, isAuth, setToken } from "../helpers";
import {
  IUserAuthenticationData,
  IUserData,
  IUserRegistrationData,
} from "../types";
import { IAPI } from "../services/API.service";
import { IPasswordData } from "../pages/Records/types";
import { IAvatarLink } from "../types/types";

export interface IUserStore {
  api: IAPI;
  theme: string | null;
  userData: IUserData;
  isAuth: boolean;
  isLoading: boolean;
  setUserData: (userData: IUserData) => void;
  setIsLoading: () => void;
  setIsAuth: () => void;
  userRegistration: (userRegistrationData: IUserRegistrationData) => void;
  userAuthentication: (userAuthenticationData: IUserAuthenticationData) => void;
  setNewUserData: (userData: IUserData) => void;
  getUserData: () => void;
  setNewPassword: (passwordData: IPasswordData) => void;
  setNewAvatar: (avatar: IAvatarLink) => void;
  setTheme: (theme: string | null) => void;
}

interface IUserAuthentication {
  token: string;
}

export class UserStore implements IUserStore {
  api: IAPI;
  theme: string | null = "light";
  userData: IUserData = {
    firstName: "",
    lastName: "",
    email: "",
    id: "",
    avatar: "",
    budgetAmount: "",
  };
  isAuth: boolean = isAuth();
  isLoading: boolean = false;

  constructor(api: IAPI) {
    makeAutoObservable(this);
    this.api = api;
  }

  setTheme = (theme: string | null) => {
    this.theme = theme;
  };

  setUserData = (userData: IUserData) => {
    this.userData = userData;
  };

  setIsLoading = () => {
    this.isLoading = !this.isLoading;
  };

  setIsAuth = () => {
    this.isAuth = isAuth();
  };

  userRegistration = async (userRegistrationData: IUserRegistrationData) => {
    try {
      this.setIsLoading();
      await this.api.fetchRequest({
        url: "/user/register",
        method: "post",
        body: userRegistrationData,
        token: null,
      });
      this.setIsAuth();
    } catch (error: any) {
      return error.response.data.details;
    } finally {
      this.setIsLoading();
    }
  };

  userAuthentication = async (
    userAuthenticationData: IUserAuthenticationData | undefined
  ) => {
    try {
      const { token }: IUserAuthentication = await this.api.fetchRequest({
        url: "/user/login",
        method: "post",
        body: userAuthenticationData,
        token: null,
      });
      setToken(token);
      this.setIsAuth();
    } catch (error: any) {
      return error.response.data.details;
    }
  };

  setNewUserData = async (userData: IUserData) => {
    try {
      this.setIsLoading();
      await this.api.fetchRequest({
        url: "/user",
        method: "patch",
        body: userData,
        token: getToken(),
      });
      await this.getUserData();
      this.setIsAuth();
    } catch (error) {
      console.error(error);
    } finally {
      this.setIsLoading();
    }
  };

  getUserData = async () => {
    try {
      const userData: IUserData = await this.api.fetchRequest({
        url: "/user",
        method: "get",
        body: null,
        token: getToken(),
      });
      this.setUserData(userData);
      return userData;
    } catch (error) {
      const userData = {
        id: "",
        lastName: "",
        firstName: "",
        avatar: "",
        email: "",
        budgetAmount: "",
      };
      return userData;
    }
  };

  setNewPassword = async (passwordData: IPasswordData) => {
    try {
      await this.api.fetchRequest({
        url: "/user/password/repeat",
        method: "patch",
        body: passwordData,
        token: getToken(),
      });
    } catch (error: any) {
      return error.response.data.details;
    }
  };

  setNewAvatar = async (avatar: IAvatarLink) => {
    try {
      await this.api.fetchRequest({
        url: "/user",
        method: "patch",
        body: avatar,
        token: getToken(),
      });
    } catch (error: any) {
      console.error(error);
    }
  };
}
