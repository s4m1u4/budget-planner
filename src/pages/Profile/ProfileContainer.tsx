import Profile from "./Profile";
import { compose } from "recompose";
import { inject, observer } from "mobx-react";
import { ProfileProps } from "./types";

export const ProfileContainer = compose<ProfileProps, {}>(
  inject(
    ({
      rootStore: {
        userStore: {
          getUserData,
          setNewUserData,
          setNewPassword,
          setNewAvatar,
        },
      },
    }) => ({
      getUserData,
      setNewUserData,
      setNewPassword,
      setNewAvatar,
    })
  ),
  observer
)(Profile);
