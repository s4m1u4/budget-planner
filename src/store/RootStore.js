import API from "../services/API";
import UserStore from "./UserStore";

export default class RootStore {
  API = new API();
  userStore = new UserStore();
}
