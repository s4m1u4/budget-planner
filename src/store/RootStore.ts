import { API, IAPI } from "services/api";
import { UserStore, IUserStore } from "store/UserStore";
import { DashboardStore, IDashboardStore } from "pages/Dashboard";

export class RootStore {
  api: IAPI = new API();
  userStore: IUserStore = new UserStore(this.api);
  dashboardStore: IDashboardStore = new DashboardStore(this.api);
}
