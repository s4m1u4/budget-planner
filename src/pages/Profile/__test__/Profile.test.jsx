import axios from "axios";
import { ProfileContainer } from "../ProfileContainer";
import { Provider } from "mobx-react";
import { render, screen } from "@testing-library/react";
import { RootStore } from "src/store/RootStore";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import userEvent from "@testing-library/user-event";

const renderWithStoreAndRouter = (initialEntries) => {
  const rootStore = new RootStore();

  return render(
    <Provider rootStore={rootStore}>
      <MemoryRouter initialEntries={initialEntries}>
        <Routes>
          <Route path="/profile" element={<ProfileContainer />} />
          <Route path="/profile/:editMode" element={<ProfileContainer />} />
        </Routes>
      </MemoryRouter>
    </Provider>
  );
};

jest.mock("axios");

describe("Profile page", () => {
  it("snapshot matching", () => {
    const { baseElement } = renderWithStoreAndRouter(["/profile"]);
    expect(baseElement).toMatchSnapshot();
  });

  describe("should render", () => {
    beforeEach(() => {
      renderWithStoreAndRouter(["/profile"]);
    });

    it("a title", async () => {
      expect(screen.getByText(/Profile page/i)).toBeInTheDocument();
    });

    it('a button "Change avatar"', () => {
      expect(
        screen.getByRole("button", { name: /change avatar/i })
      ).toBeInTheDocument();
    });
  });

  describe('modal "Set new avatar"', () => {
    beforeEach(() => {
      renderWithStoreAndRouter(["/profile"]);
    });

    it("should open", () => {
      userEvent.click(screen.getByRole("button", { name: /change avatar/i }));
      expect(screen.getByText("Set a new avatar")).toBeInTheDocument();
    });

    it("should close", () => {
      userEvent.click(screen.getByRole("button", { name: /change avatar/i }));
      expect(screen.getByText("Set a new avatar")).toBeInTheDocument();
      userEvent.click(screen.queryByText(/cancel/i));
      expect(screen.queryByText("Set a new avatar")).toBeNull();
    });
  });

  describe("should be changed", () => {
    beforeEach(() => {
      renderWithStoreAndRouter(["/profile"]);
    });

    it("to edit mode", () => {
      expect(screen.queryByText(/save/i)).not.toBeInTheDocument();
      expect(screen.queryByText(/cancel/i)).not.toBeInTheDocument();

      userEvent.click(screen.getByText(/edit/i));

      expect(screen.queryByText(/save/i)).toBeInTheDocument();
      expect(screen.queryByText(/cancel/i)).toBeInTheDocument();
    });
  });
});
