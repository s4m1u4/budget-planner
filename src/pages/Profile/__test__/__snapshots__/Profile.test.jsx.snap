// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile page snapshot matching 1`] = `
<body>
  <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div
        class="css-o6fnr7"
      >
        <h1
          class="css-15poin4"
        >
          Profile page
        </h1>
        <span
          class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-fomgco-MuiSkeleton-root"
          style="width: 200px; height: 200px;"
        />
        <div
          class="css-1o348tf"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</body>
`;
